ramping-arrival-rate: Avoid to spawn a new goroutine for every iteration

codebien opened the pull request:
Switched the `ramping-arrival-rate` from goroutine per iteration to goroutine per VU.
Refactor the Run method is not part of this PR, it is scheduled for the original issue, so I tried to impact in a minimal way the code.

However, the main issue in terms of performance was already resolved from #1955 as the following tests should assert:

using this config:

```
exports.options = {
    scenarios: {
        test1: {
            executor: 'ramping-arrival-rate',
            preAllocatedVUs: 1000,
            stages:[
                {target:100000, duration:'0s'},
                {target:100000, duration:'15s'},
            ],
        },
    },
};

exports.default =  function () {}
```

result with v0.31.1 still affected by the problem:
```
$ docker run -v $PWD:/home/k6 -i loadimpact/k6:0.31.1 run - <script.js

iterations...........: 200591 13368.444337/s
```
and the result with master
```
$ docker run -v $PWD:/home/k6 -i loadimpact/k6:master run - <script.js

iterations...........: 1499214 99925.899592/s
```

Closes #1944

<!--


  (ﾉ◕ヮ◕)ﾉ*:・ﾟ✧
  
  Thank you for your interest in contributing to the k6 project!
  
  Before you get started, we'd kindly like to ask you to read our:
    - Contribution guidelines at https://github.com/loadimpact/k6/blob/master/CONTRIBUTING.md
    - Code of Conduct at https://github.com/loadimpact/k6/blob/master/CODE_OF_CONDUCT.md
    
  Out of respect for your time, please start a discussion regarding any bigger contributions either
  in a GitHub Issue, in the community forums or in the #contributors channel of the k6 slack before you
  get started on the implementation.
  
  If you've already done all of that, you're more than welcome to proceed with your pull request.
  Thank you again for your contribution! 🙏🏼
  
  
-->

MStoykov commented:
hi @codebien , thanks for the PR :tada: !

I will try to review it later today or tomorrow, but can you post some [benchstat](https://pkg.go.dev/golang.org/x/perf/cmd/benchstat) comparisons between pre and post changes.

na-- commented on lib/executor/ramping_arrival_rate.go:
```diff
@@ -439,12 +460,13 @@ func (varr RampingArrivalRate) Run(parentCtx context.Context, out chan<- stats.S
-			go runIteration(vu) //TODO: refactor so we dont spin up a goroutine for each iteration
+		// If iterations chan has a buffered message then
+		// there aren't any active VUs free to take care of a new request.
+		if len(iterations) == 0 {
+			iterations <- struct{}{}
```
Hmm this doesn't seem right... Imagine if `len(iterations)` is `0`, but all VUs are currently busy doing some other iteration. Your code here will happily slot the new empty struct in `iterations`, but no VU will actually start executing it. Moreover, if you actually hadn't allocated all of the VUs up to `maxVUs` (i.e. if `remainingUnplannedVUs > 0`), you wouldn't start allocating a new VU here. So this iteration will be essentially stuck in limbo until a VU frees itself or a new iteration comes and we allocate a VU then.

na-- commented on lib/executor/ramping_arrival_rate.go:
```diff
@@ -316,25 +316,52 @@ func (varr RampingArrivalRate) Run(parentCtx context.Context, out chan<- stats.S
 	activeVUsCount := uint64(0)
 
 	returnVU := func(u lib.InitializedVU) {
+		exitVUIterator <- struct{}{}
```
The VU you're returning here might not be the same VU that receives the signal further down on line 360.

na-- commented on lib/executor/ramping_arrival_rate.go:
```diff
@@ -316,25 +316,52 @@ func (varr RampingArrivalRate) Run(parentCtx context.Context, out chan<- stats.S
+				select {
+				case <-iterations:
+					runIterationBasic(maxDurationCtx, activeVU)
+				case <-exitVUIterator:
```
I am not sure if you even need a new `exitVUIterator` channel. As far as I understand your code, you are using this as a condition for VUs to stop listening for new iterations. If so, listening to `regDurationCtx.Done()` should be enough, right? Read [the docs for `getDurationContexts()`](https://github.com/k6io/k6/blob/v0.31.1/lib/executor/helpers.go#L116-L141) to get an idea what the contexts are.

na-- commented on lib/executor/ramping_arrival_rate_test.go:
```diff
@@ -243,6 +243,60 @@ func TestRampingArrivalRateRunCorrectRateWithSlowRate(t *testing.T) {
+			engineOut := make(chan stats.SampleContainer, 100)
+			defer close(engineOut)
+
+			es := lib.NewExecutionState(lib.Options{}, mustNewExecutionTuple(nil, nil), 1, uint64(tc.prealloc.ValueOrZero()))
```
```suggestion
			es := lib.NewExecutionState(lib.Options{}, mustNewExecutionTuple(nil, nil), uint64(tc.prealloc.Int64), uint64(tc.prealloc.Int64))
```

Though for some reason I get a panic

na-- commented on lib/executor/ramping_arrival_rate_test.go:
```diff
@@ -243,6 +243,60 @@ func TestRampingArrivalRateRunCorrectRateWithSlowRate(t *testing.T) {
+
+	for _, tc := range tests {
+		b.Run(fmt.Sprintf("PreAllocatedVUs%d", tc.prealloc.ValueOrZero()), func(b *testing.B) {
+			engineOut := make(chan stats.SampleContainer, 100)
```
I suggest you increase the size of this buffer, as well as add a goroutine that reads from it and discards any metrics (we don't care about them in the benchmark). Otherwise, if your test emits too many `dropped_iteration` metrics, this will fill up and the test will deadlock like this:
```
$ go test -trimpath -benchmem -run='^$' -bench "BenchmarkRampingArrivalRateRun" github.com/loadimpact/k6/lib/executor

fatal error: all goroutines are asleep - deadlock!

goroutine 1 [chan receive]:
testing.(*B).run1(0xc000146900, 0xc000146900)
	testing/benchmark.go:234 +0xa5
testing.(*B).Run(0xc0001466c0, 0xae7a72, 0x1e, 0xb06678, 0x508000)
	testing/benchmark.go:667 +0x373
testing.runBenchmarks.func1(0xc0001466c0)
	testing/benchmark.go:541 +0x78
testing.(*B).runN(0xc0001466c0, 0x1)
	testing/benchmark.go:192 +0xeb
testing.runBenchmarks(0xaec487, 0x25, 0xc00000c558, 0x1043060, 0x5, 0x5, 0x104d220)
	testing/benchmark.go:550 +0x3d1
testing.(*M).Run(0xc000189880, 0x0)
	testing/testing.go:1423 +0x51d
main.main()
	_testmain.go:131 +0x138

goroutine 33 [chan receive]:
testing.(*B).run1(0xc000146b40, 0xc000146b40)
	testing/benchmark.go:234 +0xa5
testing.(*B).Run(0xc000146900, 0xc0000281b0, 0x12, 0xc000058000, 0x0)
	testing/benchmark.go:667 +0x373
github.com/loadimpact/k6/lib/executor.BenchmarkRampingArrivalRateRun(0xc000146900)
	github.com/loadimpact/k6/lib/executor/ramping_arrival_rate_test.go:255 +0x188
testing.(*B).runN(0xc000146900, 0x1)
	testing/benchmark.go:192 +0xeb
testing.(*B).run1.func1(0xc000146900)
	testing/benchmark.go:232 +0x57
created by testing.(*B).run1
	testing/benchmark.go:225 +0x7f

goroutine 9 [chan send]:
github.com/loadimpact/k6/stats.PushIfNotDone(...)
	github.com/loadimpact/k6/stats/stats.go:434
github.com/loadimpact/k6/lib/executor.RampingArrivalRate.Run(0xc0000763c0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	github.com/loadimpact/k6/lib/executor/ramping_arrival_rate.go:473 +0x1485
github.com/loadimpact/k6/lib/executor.BenchmarkRampingArrivalRateRun.func1(0xc000146b40)
	github.com/loadimpact/k6/lib/executor/ramping_arrival_rate_test.go:290 +0x62f
testing.(*B).runN(0xc000146b40, 0x1)
	testing/benchmark.go:192 +0xeb
testing.(*B).run1.func1(0xc000146b40)
	testing/benchmark.go:232 +0x57
created by testing.(*B).run1
	testing/benchmark.go:225 +0x7f

goroutine 10 [chan receive]:
github.com/loadimpact/k6/lib/executor.RampingArrivalRate.Run.func4(0xc0000322a0, 0xc0000223e8, 0xc00018a060, 0xc000032420, 0xc0000780d0, 0xba23a0, 0xc00018a000, 0xc00006c0f0)
	github.com/loadimpact/k6/lib/executor/ramping_arrival_rate.go:380 +0xc5
created by github.com/loadimpact/k6/lib/executor.RampingArrivalRate.Run
	github.com/loadimpact/k6/lib/executor/ramping_arrival_rate.go:371 +0xb29

goroutine 244 [chan send]:
github.com/loadimpact/k6/lib/executor.RampingArrivalRateConfig.cal(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	github.com/loadimpact/k6/lib/executor/ramping_arrival_rate.go:274 +0x411
created by github.com/loadimpact/k6/lib/executor.RampingArrivalRate.Run
	github.com/loadimpact/k6/lib/executor/ramping_arrival_rate.go:444 +0x10f9
```

na-- commented on lib/executor/ramping_arrival_rate.go:
```diff
@@ -316,25 +316,37 @@ func (varr RampingArrivalRate) Run(parentCtx context.Context, out chan<- stats.S
 		varr.executionState.ModCurrentlyActiveVUsCount(+1)
 		atomic.AddUint64(&activeVUsCount, 1)
+		go func() {
+			for avu := range iterations {
```
Hmm this seems a bit strange :confused: 

I haven't tested it and I haven't spotted an actual bug, but sending a VU (i.e. the _worker_) over the `iterations` (i.e. the _work_) channel is counter-intuitive. Especially when you spawn this goroutine in the context of `Activate()`-ing a specific VU, but the `avu` variable here would be assigned any random VU that is free when running the iteration.

codebien commented:
>I will try to review it later today or tomorrow, but can you post some benchstat comparisons between pre and post changes.

No improvements :worried: 

<pre>
<!doctype html>
<html>
<head>
<meta charset="utf-8">
</head>
<body>
<table class='benchstat oldnew'>
<tr class='configs'><th><th>master.bench<th>1944.bench
<tbody>
<tr><th><th colspan='2' class='metric'>iterations/s<th>delta
<tr class='better'><td>RampingArrivalRate1000VUs-8<td>1.06M ± 4%<td>0.92M ± 2%<td class='delta'>−13.61%<td class='note'>(p=0.029 n=4&#43;4)
<tr><td>&nbsp;
</tbody>
<tbody>
<tr><th><th colspan='2' class='metric'>alloc/op<th>delta
<tr class='better'><td>RampingArrivalRate1000VUs-8<td>3.99GB ± 4%<td>3.45GB ± 2%<td class='delta'>−13.60%<td class='note'>(p=0.029 n=4&#43;4)
<tr><td>&nbsp;
</tbody>
<tbody>
<tr><th><th colspan='2' class='metric'>allocs/op<th>delta
<tr class='better'><td>RampingArrivalRate1000VUs-8<td>10.6M ± 4%<td>9.2M ± 2%<td class='delta'>−13.57%<td class='note'>(p=0.029 n=4&#43;4)
<tr><td>&nbsp;
</tbody>
</table>
</body>
</html>
</pre>

```
go test -bench=BenchmarkRampingArrivalRate -run=^$ -benchmem -v ./lib/executor
BenchmarkRampingArrivalRateRun/VUs10
BenchmarkRampingArrivalRateRun/VUs10-8         	       1	    952432 iterations/s	3779520344 B/op	13655539 allocs/op
BenchmarkRampingArrivalRateRun/VUs100
BenchmarkRampingArrivalRateRun/VUs100-8        	       1	   1173044 iterations/s	4412040832 B/op	11744001 allocs/op
BenchmarkRampingArrivalRateRun/VUs1000
BenchmarkRampingArrivalRateRun/VUs1000-8       	       1	    854069 iterations/s	3213358384 B/op	 8555731 allocs/op
BenchmarkRampingArrivalRateRun/VUs10000
BenchmarkRampingArrivalRateRun/VUs10000-8      	       1	    619091 iterations/s	2343777264 B/op	 6294554 allocs/op
```

The current code adds 2xMaxVUs goroutines, a future refactor could consider re-using the goroutines spawn from `Activate`.

```
github.com/loadimpact/k6/lib/executor.RampingArrivalRate.Run.func3.1 N=1000
github.com/loadimpact/k6/lib/testutils/minirunner.(*VU).Activate.func1 N=1000 
```

na-- commented on lib/executor/ramping_arrival_rate.go:
```diff
@@ -343,13 +356,7 @@ func (varr RampingArrivalRate) Run(parentCtx context.Context, out chan<- stats.S
 	defer close(makeUnplannedVUCh)
 	go func() {
 		defer close(returnedVUs)
-		defer func() {
-			// this is done here as to not have an unplannedVU in the middle of initialization when
-			// starting to return activeVUs
-			for i := uint64(0); i < atomic.LoadUint64(&activeVUsCount); i++ {
-				<-activeVUs
-			}
-		}()
```
I understand why this was deleted, but I don't see what replaces it? What is ensuring that all iterations are done before the executor is done and before the VUs are returned to the [common VU buffer pool in the `ExecutionState`](https://github.com/k6io/k6/blob/v0.31.1/lib/execution.go#L177)?

na-- replied:
Also, this has the effect of ignoring `gracefulStop`, try running the following script with your branch and with the latest k6 `master` (or the official release):
```js
import { sleep } from 'k6';

export let options = {
    scenarios: {
        rarr: {
            executor: 'ramping-arrival-rate',
            startRate: 10, timeUnit: '1s',
            stages: [
                { target: 10, duration: '5s' },
            ],
            preAllocatedVUs: 10,
            maxVUs: 10,
            gracefulStop: '30s',
        },
    }
}

export default function () {
    sleep(10); // longer than the duration, but shorter than duration+gracefulStop
    console.log(`VU ${__VU} finished iteration ${__ITER}`)
}
```

With the k6 `master` code it will wait for the first 10 iterations to finish, because of the `gracefulStop`, and you'd get something like this:
```
  scenarios: (100.00%) 1 scenario, 10 max VUs, 35s max duration (incl. graceful stop):
           * rarr: Up to 10.00 iterations/s for 5s over 1 stages (maxVUs: 10, gracefulStop: 30s)

WARN[0001] Insufficient VUs, reached 10 active VUs and cannot initialize more  executor=ramping-arrival-rate scenario=rarr
INFO[0010] VU 2 finished iteration 0                     source=console
INFO[0010] VU 4 finished iteration 0                     source=console
INFO[0010] VU 3 finished iteration 0                     source=console
INFO[0010] VU 5 finished iteration 0                     source=console
INFO[0010] VU 6 finished iteration 0                     source=console
INFO[0011] VU 7 finished iteration 0                     source=console
INFO[0011] VU 8 finished iteration 0                     source=console
INFO[0011] VU 9 finished iteration 0                     source=console
INFO[0011] VU 10 finished iteration 0                    source=console
INFO[0011] VU 1 finished iteration 0                     source=console

running (11.0s), 00/10 VUs, 10 complete and 0 interrupted iterations
rarr ✓ [======================================] 10/10 VUs  5s  10 iters/s

     data_received........: 0 B 0 B/s
     data_sent............: 0 B 0 B/s
     dropped_iterations...: 39  3.544626/s
     iteration_duration...: avg=10s min=10s med=10s max=10s p(90)=10s p(95)=10s
     iterations...........: 10  0.908878/s
     vus..................: 10  min=10 max=10
     vus_max..............: 10  min=10 max=10
```

with your branch, I get:
```
  scenarios: (100.00%) 1 scenario, 10 max VUs, 35s max duration (incl. graceful stop):
           * rarr: Up to 10.00 iterations/s for 5s over 1 stages (maxVUs: 10, gracefulStop: 30s)

WARN[0001] Insufficient VUs, reached 10 active VUs and cannot initialize more  executor=ramping-arrival-rate scenario=rarr

running (05.0s), 00/10 VUs, 0 complete and 10 interrupted iterations
rarr ✓ [======================================] 08/10 VUs  5s  10 iters/s
WARN[0005] No script iterations finished, consider making the test duration longer 

     data_received........: 0 B 0 B/s
     data_sent............: 0 B 0 B/s
     dropped_iterations...: 39  7.798556/s
     vus..................: 10  min=10 max=10
     vus_max..............: 10  min=10 max=10
```

na-- reviewed (changes requested):
I noted the problem in an inline comment, but to re-iterate, this change brakes `gracefulStop` and the assurance that VUs are done executing JS when they are returned to the common pool, which may cause a data race if we try to run another scenario on the same VU (JS runtimes are single-threaded).

codebien commented:
Added a test to cover the graceful case, I polished a bit the solution fixing the graceful issue.

na-- commented on lib/executor/ramping_arrival_rate.go:
```diff
@@ -471,3 +467,49 @@ func (varr RampingArrivalRate) Run(parentCtx context.Context, out chan<- stats.S
 	return nil
 }
+
+// runningVUs controls the activeVUs
+// executing the received requests for iterations.
+type runningVUs struct {
+	Iterations chan struct{}
```
I :heart:  the approach of splitting this into a separate small component, though I'm not 100% sure about the names... Though my ideas aren't necessarily any better... :sweat_smile: `VUPool`? `ActiveVUPool`? 

In any case, `Iterations` probably should be not be exported and the comment looks a bit wonky.

imiric replied:
`activeVUPool` sounds slightly better to me.

Also `TryRunIteration` or `RunIterationMaybe` instead of `Run`. Not sure about the `Try`/`Maybe`, but `Iteration` should be part of it.

na-- reviewed (approved):
LGTM besides some minor naming nitpicks (for which I don't really have any better ideas :sweat_smile:)

MStoykov commented on lib/executor/ramping_arrival_rate.go:
```diff
@@ -316,25 +316,36 @@ func (varr RampingArrivalRate) Run(parentCtx context.Context, out chan<- stats.S
 	returnedVUs := make(chan struct{})
 	startTime, maxDurationCtx, regDurationCtx, cancel := getDurationContexts(parentCtx, duration, gracefulStop)
 
+	iterations := make(chan struct{})
+	iterators := runningVUs{
+		Iterations: iterations,
+	}
+
 	defer func() {
+		close(iterations)
```
I would prefer that you encapsualte the `iterations` as well and have `newRunningVUs` (or IMO `newVUPool`) which make the `iterations` channel and a method that is `stop` to close `iterations` and effectively stop the "running" of vus.

MStoykov reviewed (commented):
LGTM in general, good job! 

I have some nitpicks, one of which grew while I was writing it out :sweat_smile: , but nothing blocking IMO ... but we are in code freeze so ;)

imiric commented on lib/executor/ramping_arrival_rate_test.go:
```diff
@@ -243,6 +243,109 @@ func TestRampingArrivalRateRunCorrectRateWithSlowRate(t *testing.T) {
+			PreAllocatedVUs: null.IntFrom(10),
+			MaxVUs:          null.IntFrom(10),
+			BaseConfig: BaseConfig{
+				GracefulStop: types.NullDurationFrom(15 * time.Second),
```
Nit: you could reduce this duration and the sleep to speed the test up, there's no reason for it to take 5s+. Even though it's not noticeable when running the full suite, it matters when running it standalone.

imiric reviewed (commented):
LGTM, don't have much to add besides the naming issues already mentioned, and that all `runningVUs` fields/methods don't need to be exported.

Even though the main performance issue was addressed in #1955, from a quick profile run this does get rid of the excessive stack creation mentioned in https://github.com/k6io/k6/issues/1386#issuecomment-780704639, so definitely an improvement. :+1:

codebien commented:
Done and Rebased

Renamed the struct ad the run method, unexported `iterations` chan and reduced the time for the graceful option in the relative test.

imiric commented on lib/executor/ramping_arrival_rate.go:
```diff
@@ -471,3 +463,57 @@ func (varr RampingArrivalRate) Run(parentCtx context.Context, out chan<- stats.S
+	}
+}
+
+// Run invokes a request to execute a new iteration.
+// When there aren'p *available VUs to process the request
```
```suggestion
// TryRunIteration invokes a request to execute a new iteration.
// When there are no available VUs to process the request
```

We might also unexport all the `activeVUPool` methods, but it's not a big deal.

imiric commented on lib/executor/ramping_arrival_rate.go:
```diff
@@ -471,3 +463,57 @@ func (varr RampingArrivalRate) Run(parentCtx context.Context, out chan<- stats.S
+}
+
+// Close stops the pool from accepting requests
+// then it will wait thap *all the on-going iterations will be completed.
```
```suggestion
// then it will wait for all on-going iterations to complete.
```

imiric reviewed (approved):
LGTM, just noticed some search/replace typos :)

MStoykov commented on lib/executor/ramping_arrival_rate.go:
```diff
@@ -316,25 +316,32 @@ func (varr RampingArrivalRate) Run(parentCtx context.Context, out chan<- stats.S
+	vusPool := newActiveVUPool()
+
 	defer func() {
+		vusPool.Close()
 		// Make sure all VUs aren't executing iterations anymore, for the cancel()
 		// below to deactivate them.
 		<-returnedVUs
 		cancel()
 		activeVUsWg.Wait()
 	}()
```
This should prevent the race from `https://github.com/k6io/k6/pull/1957/checks?check_run_id=2631773502`
```
==================
WARNING: DATA RACE
Read at 0x00c000062078 by goroutine 128:
  internal/race.Read()
      /opt/hostedtoolcache/go/1.16.4/x64/src/internal/race/race.go:37 +0x206
  sync.(*WaitGroup).Add()
      /opt/hostedtoolcache/go/1.16.4/x64/src/sync/waitgroup.go:71 +0x219
  go.k6.io/k6/lib/executor.(*activeVUPool).AddVU()
      /home/runner/work/k6/k6/lib/executor/ramping_arrival_rate.go:519 +0x64
  go.k6.io/k6/lib/executor.RampingArrivalRate.Run.func3()
      /home/runner/work/k6/k6/lib/executor/ramping_arrival_rate.go:361 +0x290
  go.k6.io/k6/lib/executor.RampingArrivalRate.Run.func4()
      /home/runner/work/k6/k6/lib/executor/ramping_arrival_rate.go:379 +0x571

Previous write at 0x00c000062078 by goroutine 34:
  internal/race.Write()
      /opt/hostedtoolcache/go/1.16.4/x64/src/internal/race/race.go:41 +0x125
  sync.(*WaitGroup).Wait()
      /opt/hostedtoolcache/go/1.16.4/x64/src/sync/waitgroup.go:128 +0x126
  go.k6.io/k6/lib/executor.(*activeVUPool).Close()
      /home/runner/work/k6/k6/lib/executor/ramping_arrival_rate.go:535 +0x77
  go.k6.io/k6/lib/executor.RampingArrivalRate.Run.func1()
      /home/runner/work/k6/k6/lib/executor/ramping_arrival_rate.go:339 +0x54
  go.k6.io/k6/lib/executor.RampingArrivalRate.Run()
      /home/runner/work/k6/k6/lib/executor/ramping_arrival_rate.go:481 +0x20d0
  go.k6.io/k6/lib/executor.(*RampingArrivalRate).Run()
      <autogenerated>:1 +0x127
  go.k6.io/k6/lib/executor.TestRampingArrivalRateRunUnplannedVUs()
      /home/runner/work/k6/k6/lib/executor/ramping_arrival_rate_test.go:189 +0x7ee
  testing.tRunner()
      /opt/hostedtoolcache/go/1.16.4/x64/src/testing/testing.go:1193 +0x202
```

As far as I can see the waitgroup was empty at the time as a VU wasn't executing but it was *just* starting to and it raced between the `Add` and the `Close`. This change makes certain we no longer *will* try to "Add" before we close 
```suggestion
		// Make sure all VUs aren't executing iterations anymore, for the cancel()
		// below to deactivate them.
		<-returnedVUs
		cancel()
		vusPool.Close()
		activeVUsWg.Wait()
	}()
```
